carvel-packages/installer/bundle/config/ytt/_ytt_lib/config/functions.star

```diff
@@ -0,0 +1,17 @@
+load("@ytt:struct", "struct")
+
+def removeNulls(data):
+   # Iterate over a struct of scalar values and return only those where value is not null
+   filtered_data = {}
+   for key in struct.decode(data):
+      value = getattr(data, key, None)
+      if type(value) == "struct":
+         value = removeNulls(value)
+      end
+      if value: #! This means that value is not an empty string, dict, struct, ...
```

Comment:
Will also remove empty lists and dicts. Is that always a good idea???

Reply:
DIscussed in slack. This is removing empty lists and empty dicts in the provided config that is recorded in the secret. Typically this is fine, as user will most likely never provide empty lists or empty dicts and these are generated by the ytt schema defaulting things. Some are not a big issue, but many in the main `educates` top level config are really ugly to see. e.g.:
```
    debug: false
    clusterPackages:
      contour:
        enabled: true
        settings:
          infraProvider: kind
          contour:
            replicas: 1
          configFileContents:
            defaultHttpVersions:
            - HTTP/1.1
          service:
            type: ClusterIP
            useHostPorts: true
      cert-manager:
        enabled: false
        settings: {}
      external-dns:
        enabled: false
        settings: {}
      certs:
        enabled: false
        settings: {}
      kyverno:
        enabled: true
        settings: {}
      kapp-controller:
        enabled: false
        settings: {}
      educates:
        enabled: false
        settings:
          imageVersions:
          - name: session-manager
            image: ghcr.io/vmware-tanzu-labs/educates-session-manager:3.0.0-alpha.4
          - name: training-portal
            image: ghcr.io/vmware-tanzu-labs/educates-training-portal:3.0.0-alpha.4
          - name: docker-registry
            image: ghcr.io/vmware-tanzu-labs/educates-docker-registry:3.0.0-alpha.4
          - name: pause-container
            image: ghcr.io/vmware-tanzu-labs/educates-pause-container:3.0.0-alpha.4
          - name: base-environment
            image: ghcr.io/vmware-tanzu-labs/educates-base-environment:3.0.0-alpha.4
          - name: jdk8-environment
            image: ghcr.io/vmware-tanzu-labs/educates-jdk8-environment:3.0.0-alpha.4
          - name: jdk11-environment
            image: ghcr.io/vmware-tanzu-labs/educates-jdk11-environment:3.0.0-alpha.4
          - name: jdk17-environment
            image: ghcr.io/vmware-tanzu-labs/educates-jdk17-environment:3.0.0-alpha.4
          - name: jdk21-environment
            image: ghcr.io/vmware-tanzu-labs/educates-jdk21-environment:3.0.0-alpha.4
          - name: conda-environment
            image: ghcr.io/vmware-tanzu-labs/educates-conda-environment:3.0.0-alpha.4
          - name: secrets-manager
            image: ghcr.io/vmware-tanzu-labs/educates-secrets-manager:3.0.0-alpha.4
          - name: tunnel-manager
            image: ghcr.io/vmware-tanzu-labs/educates-tunnel-manager:3.0.0-alpha.4
          - name: image-cache
            image: ghcr.io/vmware-tanzu-labs/educates-image-cache:3.0.0-alpha.4
          - name: assets-server
            image: ghcr.io/vmware-tanzu-labs/educates-assets-server:3.0.0-alpha.4
          - name: debian-base-image
            image: debian:sid-20230502-slim
          - name: docker-in-docker
            image: docker:20.10.18-dind
          - name: rancher-k3s-v1.27
            image: rancher/k3s:v1.27.14-k3s1
          - name: rancher-k3s-v1.28
            image: rancher/k3s:v1.28.10-k3s1
          - name: rancher-k3s-v1.29
            image: rancher/k3s:v1.29.5-k3s1
          - name: rancher-k3s-v1.30
            image: rancher/k3s:v1.30.1-k3s1
          - name: loftsh-vcluster
            image: loftsh/vcluster:0.18.1
          clusterIngress:
            domain: kind.mac16.tanzu-devs.com
          clusterSecurity:
            policyEngine: kyverno
          workshopSecurity:
            rulesEngine: kyverno
    clusterInfrastructure:
      provider: kind
      aws: null
      gcp: null
      caCertificateRef: null
    localKindCluster:
      listenAddress: ""
      apiServer:
        address: ""
        port: 6443
      volumeMounts: []
    localDNSResolver:
      targetAddress: ""
      extraDomains: []
    clusterIngress:
      domain: null
      class: null
      protocol: null
      tlsCertificate: null
      tlsCertificateRef: null
      caCertificate: null
      caCertificateRef: null
      caNodeInjector: null
    clusterSecurity:
      policyEngine: kyverno
    workshopSecurity:
      rulesEngine: kyverno
```

I could modify it so that only nulls things that are not under clusterPackages, which is the most relevant part, although the fact that they are removed don't have any influence.

I would like to be able to record the values the user provides, but as these are processed by the ytt schema, the defaults are provided, and make things less readable for all the `#@schema/nullable` objects

Reply:
We can always modify line 8 of functions.star to do:
```
      if type(value) == "struct" and key != "clusterPackages":
```
This way at least the clusterPackages keep the empty settings if they are provided.